arrow-libs/core/arrow-core-data/src/main/kotlin/arrow/core/NonEmptyList.kt

```diff
@@ -550,8 +429,165 @@ object NonEmptyListSemigroup : Semigroup<NonEmptyList<Any?>> {
+  zip(
+    b, c, d,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit
+  ) { aa, bb, cc, dd, _, _, _, _, _, _ ->
+    map(aa, bb, cc, dd)
+  }
+
+inline fun <A, B, C, D, E, Z> NonEmptyList<A>.zip(
+  b: NonEmptyList<B>,
+  c: NonEmptyList<C>,
+  d: NonEmptyList<D>,
+  e: NonEmptyList<E>,
+  map: (A, B, C, D, E) -> Z
+): NonEmptyList<Z> =
+  zip(
+    b, c, d, e,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit
+  ) { aa, bb, cc, dd, ee, _, _, _, _, _ ->
+    map(aa, bb, cc, dd, ee)
+  }
+
+inline fun <A, B, C, D, E, F, Z> NonEmptyList<A>.zip(
+  b: NonEmptyList<B>,
+  c: NonEmptyList<C>,
+  d: NonEmptyList<D>,
+  e: NonEmptyList<E>,
+  f: NonEmptyList<F>,
+  map: (A, B, C, D, E, F) -> Z
+): NonEmptyList<Z> =
+  zip(
+    b, c, d, e, f,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit
+  ) { aa, bb, cc, dd, ee, ff, _, _, _, _ ->
+    map(aa, bb, cc, dd, ee, ff)
+  }
+
+inline fun <A, B, C, D, E, F, G, Z> NonEmptyList<A>.zip(
+  b: NonEmptyList<B>,
+  c: NonEmptyList<C>,
+  d: NonEmptyList<D>,
+  e: NonEmptyList<E>,
+  f: NonEmptyList<F>,
+  g: NonEmptyList<G>,
+  map: (A, B, C, D, E, F, G) -> Z
+): NonEmptyList<Z> =
+  zip(
+    b, c, d, e, f, g,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit
+  ) { aa, bb, cc, dd, ee, ff, gg, _, _, _ ->
+    map(aa, bb, cc, dd, ee, ff, gg)
+  }
+
+inline fun <A, B, C, D, E, F, G, H, Z> NonEmptyList<A>.zip(
+  b: NonEmptyList<B>,
+  c: NonEmptyList<C>,
+  d: NonEmptyList<D>,
+  e: NonEmptyList<E>,
+  f: NonEmptyList<F>,
+  g: NonEmptyList<G>,
+  h: NonEmptyList<H>,
+  map: (A, B, C, D, E, F, G, H) -> Z
+): NonEmptyList<Z> =
+  zip(
+    b, c, d, e, f, g, h,
+    NonEmptyList.unit, NonEmptyList.unit
+  ) { aa, bb, cc, dd, ee, ff, gg, hh, _, _ ->
+    map(aa, bb, cc, dd, ee, ff, gg, hh)
+  }
+
+inline fun <A, B, C, D, E, F, G, H, I, Z> NonEmptyList<A>.zip(
+  b: NonEmptyList<B>,
+  c: NonEmptyList<C>,
+  d: NonEmptyList<D>,
+  e: NonEmptyList<E>,
+  f: NonEmptyList<F>,
+  g: NonEmptyList<G>,
+  h: NonEmptyList<H>,
+  i: NonEmptyList<I>,
+  map: (A, B, C, D, E, F, G, H, I) -> Z
+): NonEmptyList<Z> =
+  zip(
+    b, c, d, e, f, g, h, i,
+    NonEmptyList.unit
+  ) { aa, bb, cc, dd, ee, ff, gg, hh, ii, _ ->
+    map(aa, bb, cc, dd, ee, ff, gg, hh, ii)
+  }
+
+inline fun <A, B, C, D, E, F, G, H, I, J, Z> NonEmptyList<A>.zip(
+  b: NonEmptyList<B>,
+  c: NonEmptyList<C>,
+  d: NonEmptyList<D>,
+  e: NonEmptyList<E>,
+  f: NonEmptyList<F>,
+  g: NonEmptyList<G>,
+  h: NonEmptyList<H>,
+  i: NonEmptyList<I>,
+  j: NonEmptyList<J>,
+  map: (A, B, C, D, E, F, G, H, I, J) -> Z
+): NonEmptyList<Z> =
+  NonEmptyList(
+    map(head, b.head, c.head, d.head, e.head, f.head, g.head, h.head, i.head, j.head),
+    tail.zip(b.tail, c.tail, d.tail, e.tail, f.tail, g.tail, h.tail, i.tail, j.tail, map)
```

Comment:
I think this is right, but will probably not work still, we might need to update the `unit` variable